src/main/java/com/snowflake/kafka/connector/SnowflakeSinkTaskAuthorizationExceptionTracker.java

```diff
@@ -0,0 +1,55 @@
+package com.snowflake.kafka.connector;
+
+import static com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig.ENABLE_TASK_FAIL_ON_AUTHORIZATION_ERRORS;
+import static com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig.ENABLE_TASK_FAIL_ON_AUTHORIZATION_ERRORS_DEFAULT;
+import static com.snowflake.kafka.connector.internal.SnowflakeErrors.ERROR_1005;
+
+import java.util.Map;
+
+/**
+ * When the user rotates Snowflake key that is stored in an external file the Connector hangs and
+ * does not mark its tasks as failed. To fix this corner case we need to track the authorization
+ * exception thrown during preCommit() and stop tasks during put().
+ *
+ * <p>Note that exceptions thrown during preCommit() are swallowed by Kafka Connect and will not
+ * cause task failure.
+ */
+public class SnowflakeSinkTaskAuthorizationExceptionTracker {
+
+  public static final String AUTHORIZATION_EXCEPTION_MESSAGE = "Authorization failed after retry";
+
+  boolean authorizationTaskFailureEnabled;
+  boolean authorizationErrorReported;
+
+  public SnowflakeSinkTaskAuthorizationExceptionTracker() {
+    this.authorizationTaskFailureEnabled = true;
+    this.authorizationErrorReported = false;
+  }
+
+  public void updateStateOnTaskStart(Map<String, String> taskConfig) {
+    authorizationTaskFailureEnabled =
+        Boolean.parseBoolean(
+            taskConfig.getOrDefault(
+                ENABLE_TASK_FAIL_ON_AUTHORIZATION_ERRORS,
+                Boolean.toString(ENABLE_TASK_FAIL_ON_AUTHORIZATION_ERRORS_DEFAULT)));
+  }
+
+  /**
+   * Check if the thrown exception is related to authorization
+   *
+   * @param ex - any exception that occurred during preCommit
+   */
+  public void reportPrecommitException(Exception ex) {
+    if (authorizationTaskFailureEnabled
+        && ex.getMessage().contains(AUTHORIZATION_EXCEPTION_MESSAGE)) {
+      authorizationErrorReported = true;
+    }
+  }
+
+  /** Throw exception if authorization has failed before */
+  public void throwExceptionIfAuthorizationFailed() {
+    if (authorizationTaskFailureEnabled && authorizationErrorReported) {
+      throw ERROR_1005.getException();
+    }
+  }
```

Comment:
I think that in `reportPrecommitException()` we do not need a feature check. toggling a field is not dangerous and gives us another possibility...

...in `throwExceptionIfAuthorizationFailed()` if the feature is disabled by default (can we check if it is default vs explicitly disabled?), we could add a warning log — something like:
> Authorization problem detected. You may want to configure "enable.task.fail.on.authorization.errors = true" to make the connector automatically pickup rotated credentials

It could help users with discoverability and fixing problem on their own — WDYT?

Reply:
I think it might not be the best idea because of the following reasons:
- failing task doesn't mean that it's going to be restarted. This behaviour is specific to Confluent platform.
- authorization failure might happend from different reasons than key rotation
- there would be a lot of this log as they would be produced on every put() call

Overall it's a good idea, but I have doubts about this specific scenario.

Reply:
What you have just wrote could be a very fitting warning message then. We could create a latch to print it only once. Food for thought. Anyway — good job :)